connectors/services/job_executor.py

```diff
@@ -0,0 +1,228 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+"""
+Event loop
+
+- polls for work by calling Elasticsearch on a regular basis
+- instantiates connector plugins
+- mirrors an Elasticsearch index with a collection of documents
+"""
+import asyncio
+import os
+import time
+
+from connectors.byoc import BYOIndex, DataSourceError, PipelineSettings, SyncJobIndex
+from connectors.byoei import ElasticServer
+from connectors.es import defaults_for
+from connectors.logger import logger
+from connectors.services.base import BaseService
+from connectors.source import DataSourceConfiguration, get_source_klass
+
+
+class SyncJobRunner:
+    def __init__(
+        self,
+        sources,
+        sync_job,
+        elastic_server,
+        connector_index,
+        sync_job_index,
+        bulk_options,
+    ):
+        self.sources = sources
+        self.sync_job = sync_job
+        self.elastic_server = elastic_server
+        self.bulk_options = bulk_options
+        self.connector_index = connector_index
+        self.sync_job_index = sync_job_index
+
+    async def execute(self):
+        await self.sync_job.claim()
+
+        start_time = time.time()
+        index_name = self.sync_job.connector["index_name"]
```

Comment:
Performance nit

You are using `self.sync_job.connector` everywhere in this function. This adds 3 indirections in Python.

It's best to do:

```
connector = self.sync_job.connector
service_type = connector["service_type"]
language_code = connector["language"]
```

Reply:
Probably we could have this in `SyncJob` class:

```python
class SyncJob:

    ...

    @property
    def connector():
      self.doc_source.get('connector', {})

    @property
    def service_type():
      self.connector['service_type']

    @property
    def language_code():
      self.connector['language']
```

cc. @vidok

Reply:
well it's not solving the performance issue hehe, but it's cleaner, I like it!